### PR TITLE
LDAP UserAccountControl support

### PR DESCRIPTION
**Pull request by yaleman**
Per discussions offline, the LDAP UserAccountControl field is a bitfield of flags, which doesn't have a direct mapping back to kanidm things. This PR adds two things

 - a temporary hack to respond with the gidnumber to allow queries while I'm working on other things
 - an enum which allows for flags to be built
 
- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes

### CODE REVIEW

File: kanidmd/src/lib/constants/ldap.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    ErrorPasswordExpired,
+    ///  The account is enabled for delegation. This is a security-sensitive setting; accounts with this option enabled SHOULD be strictly controlled. This setting enables a service running under the account to assume a client identity and authenticate as that user to other remote servers on the network.
+    TrustedToAuthnForDelegation,
+    ///  (Windows Server 2008/Windows Server 2008 R2) The account is a Read-Only Domain Controller (RODC). This is a security-sensitive setting. Removing this setting from an RODC compromises security on that server.
+    PartialSecretsAccount,
+    ///  Restrict this UserPrincipalName to use only Advanced Encryption Standard (AES) encryption types for keys. This bit is ignored by Windows Client and Windows Servers.
+    UserUseAesKeys,
+}
+
+impl From<LdapUacFlag> for u32 {
```
Comment (Firstyear) [inline kanidmd/src/lib/constants/ldap.rs:69]: I think you could do LdapUacFlag as repr=u32 instead here?

File: kanidmd/src/lib/constants/ldap.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
+//! Constants for the LDAP server
+
```
Comment (Firstyear) [inline kanidmd/src/lib/constants/ldap.rs:2]: This could be in https://github.com/kanidm/ldap3_server potentially?

File: kanidmd/src/lib/ldap.rs [hunk 3]
```diff
@@ -529,6 +530,8 @@ pub(crate) fn ldap_vattr_map(input: &str) -> &str {
         "sshpublickey" => "ssh_publickey",
         "cn" => "name",
         "uidnumber" => "gidnumber",
+        // TODO: temporary hack to make it work until the [kanidm::constants::ldap::LdapUacFlag] enum can be turned into a filter
```
Comment (Firstyear) [inline kanidmd/src/lib/ldap.rs:533]: Why can't it be a filter right now?

File: kanidmd/src/lib/schema.rs [hunk 4]
```diff
@@ -1121,6 +1121,19 @@ impl<'a> SchemaWriteTransaction<'a> {
                     syntax: SyntaxType::UINT32,
                 },
             );
+            self.attributes.insert(
+                AttrString::from("useraccountcontrol"),
+                SchemaAttribute {
+                    name: AttrString::from("useraccountcontrol"),
+                    uuid: *UUID_SCHEMA_ATTR_USERACCOUNTCONTROL,
+                    description: String::from("An LDAP Compatible UserAccountControl"),
```
Comment (Firstyear) [inline kanidmd/src/lib/schema.rs:1129]: Probably should say it's MS-AD compat

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: kanidmd/src/lib/ldap.rs (hunk 3 above)
Comment (yaleman) [reply kanidmd/src/lib/ldap.rs:533]: Because I don't know how to write them? 😅

File: kanidmd/src/lib/constants/ldap.rs (hunk 2 above)
Comment (yaleman) [reply kanidmd/src/lib/constants/ldap.rs:2]: It sure could, I didn't realise it was its own crate :)

File: kanidmd/src/lib/constants/ldap.rs (hunk 1 above)
Comment (yaleman) [reply kanidmd/src/lib/constants/ldap.rs:69]: I really don't know enough rust to know what you mean by that, and google's not helping sorry :)

File: kanidmd/src/lib/schema.rs (hunk 4 above)
Comment (yaleman) [reply kanidmd/src/lib/schema.rs:1129]: Done

File: kanidmd/src/lib/constants/ldap.rs (hunk 1 above)
Comment (yaleman) [reply kanidmd/src/lib/constants/ldap.rs:76]: .. fixed :D

File: kanidmd/src/lib/ldap.rs (hunk 3 above)
Comment (Firstyear) [reply kanidmd/src/lib/ldap.rs:533]: Isn't it the other way here? We need to synthesise the UACflags from other values?

File: kanidmd/src/lib/constants/ldap.rs (hunk 2 above)
Comment (Firstyear) [reply kanidmd/src/lib/constants/ldap.rs:2]: For now we can do it here until it works, then we can add this to ldap3_server later :)

File: kanidmd/src/lib/ldap.rs (hunk 3 above)
Comment (yaleman) [reply kanidmd/src/lib/ldap.rs:533]: Yes, I don't know the right words for the thing, I just put in a hack to make it work for testing 😄  You mentioned elsewhere it'd need to be... oh, a plugin. 

Words are hard. Derp.

**Comment by yaleman** [comment]
No longer needed

**Pull request closed without merging**